### PR TITLE
[Consensus] synchronize blocks via commit certificate

### PR DESCRIPTION
**Pull request by mwtian**
## Description 

A node can miss recent blocks sent by peers because of downtime or network jitter. When too many rounds of blocks are missed, it can be hard for the node to catchup with block synchronization which is not optimized for throughput: only 1-2 level of ancestor blocks are fetched in each batch, and processing of blocks can only start after the causal history is totally complete. In experiments, we see less blocks fetched in block synchronization compared to the block production rate which can be ~1000/s.

`CommitSyncer` from this PR has the following optimizations for throughput of block fetching and processing:
- Fetching commits by range, and in parallel against multiple peers.
- Minimizing verifications, by verifying the last fetched commit with blocks, verifying other commits with previous commit digests, and verifying blocks based on the references from commits.

Fetched blocks are sent to Core for processing, which simplifies the implementation. In future we can investigate directly accepting certified blocks and sending certified commits to execution. In private testnet, block processing rate during catchup can already be 6x - 20x of normal block production rate, under 5K TPS.

## Test plan 

CI
Private testnet for catchup throughput testing, with 5K TPS

---

## Release notes

Check each box that your changes affect. If none of the boxes relate to your changes, release notes aren't required.

For each box you select, include information after the relevant heading that describes the impact of your changes that a user might notice and any actions they must take to implement updates. 

- [ ] Protocol: 
- [ ] Nodes (Validators and Full nodes): 
- [ ] Indexer: 
- [ ] JSON-RPC: 
- [ ] GraphQL: 
- [ ] CLI: 
- [ ] Rust SDK:

### CODE REVIEW

File: consensus/core/src/commit_syncer.rs
Comment (arun-koshy) [inline consensus/core/src/commit_syncer.rs:531]: s/consequtive/consecutive

File: consensus/core/src/commit_syncer.rs
Comment (arun-koshy) [inline consensus/core/src/commit_syncer.rs:538]: Add a todo to use scores here? If you agree we can use the scores to fetch from "good" hosts instead of all available hosts. We can do this in live syncer as well

File: consensus/core/src/commit_syncer.rs
Comment (arun-koshy) [inline consensus/core/src/commit_syncer.rs:331]: I think I am misunderstanding or missing something. It seems like we are getting the blocks & commits in the `fetch_commits` call and then verifying the blocks in `verify_commits`. But then we are fetching the same blocks again and doing some light digest verification? If that is true, can you explain why we need to do it twice?

File: consensus/core/src/commit_syncer.rs
Comment (arun-koshy) [inline consensus/core/src/commit_syncer.rs:336]: s/reqeust_block_refs/request_block_refs

File: consensus/core/src/network/anemo_network.rs
Comment (arun-koshy) [inline consensus/core/src/network/anemo_network.rs:593]: nit: `signed_blocks`?

File: consensus/core/src/network/tonic_network.rs
Comment (arun-koshy) [inline consensus/core/src/network/tonic_network.rs:662]: nit: same as anemo, `signed_blocks`?

File: consensus/core/src/authority_service.rs
Comment (mwtian) [reply consensus/core/src/authority_service.rs:233]: Good suggestion.

File: consensus/core/src/network/anemo_network.rs
Comment (mwtian) [reply consensus/core/src/network/anemo_network.rs:593]: I want to make it clear that the purpose of the blocks are to certify commits.

File: consensus/core/src/network/tonic_network.rs
Comment (mwtian) [reply consensus/core/src/network/tonic_network.rs:662]: Similarly I want to indicate that the blocks are for certifying the commits.

File: consensus/config/src/parameters.rs
Comment (mwtian) [reply consensus/config/src/parameters.rs]: This is indeed confusing. Updated comment for the recommended range of `commit_sync_batch_size`.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:129]: Yeah `start` and `end` should be inclusive. Comments and logs are fixed.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:157]: Good point, updated.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:121]: Since this is taking max, the result should be the same?

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs]: Updated the comments in a few places and let me know where else they can be improved. The returned blocks from `fetch_commit()` RPC only contains blocks certifying the last returned commit. They are not the blocks referenced by the commit. A separate step is needed to fetch the blocks referenced by the commits. Since the commits are certified, the referenced blocks can be considered verified as well.

The reasons for the separate block fetching step are:
1. Blocks can be large, so it seems better to reuse the chunking & size limiting logic in `fetch_blocks()`.
2. In future, we can implement logic where after commits are successfully fetched and certified, only block fetching are retried if failed.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:567]: Added to the comment above `FetchState.available_authorities`. This component needs to balance load among as many peers as possible, while avoid unhealthy peers.

File: consensus/core/src/commit_syncer.rs
Comment (akichidis) [inline consensus/core/src/commit_syncer.rs:561]: As we talked offline maybe can change the order here and do something like `BTreeSet<(Instant, u32, AuthorityIndex)>` so the authorities with the lowest number of errors are prioritised?

**Review by akichidis — APPROVED** [review]
Great work @mwtian ! I am approving to unblock but I would really like some test coverage on the commit syncer - either on this or follow up PR. There is delicate logic and having some verification to begin with would be necessary.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs]: Yes it is indeed possible to have blocks suspended here right now. When a validator equivocates, some blocks from the validator may never make it into a commit because only blocks higher than the last commit round of the authority get into new commits. But the block can still be an ancestor to other committed blocks. We need to rely on block sync here instead.

In future maybe we can implement a form of GC to avoid suspending blocks in these cases when ancestors cannot be committed. But it will take a bit of thinking.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs]: Good suggestion! Using `BTreeSet` is indeed more robust. I may wait for Arun's #17026 to land to use `CommitRange` instead.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs]: Yes this is intentional, to avoid generating a lot of small batches to fetch, which will burden this node and its peers. Suppose the node falls behind by 1 commit every now and then, if the smaller batch that includes the highest_valid_index gets generated, there will be a lot of 1 commit fetches. The existing logic will wait for the node to fall behind by 100 commit, then create a bigger batch. Expanded the comment.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:411]: Yeah agree that will be easier to write, but indeed I don't like creating VerifiedBlock before the block is actually verified. Let's revisit this in future. Maybe we can improve BlockRef and SignedBlock API around this code path.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:602]: Yeah there is one partially done in AuthorityNode. Let me add a TODO.

File: consensus/core/src/commit_syncer.rs
Comment (mwtian) [reply consensus/core/src/commit_syncer.rs:185]: Added `commit_sync_total_fetched_blocks_size` above.

**Pull request merged**
